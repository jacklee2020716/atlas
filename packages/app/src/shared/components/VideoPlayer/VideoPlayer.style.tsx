import { css } from '@emotion/core'
import { breakpoints } from '../../theme'

export type VideoPlayerStyleProps = {
  width?: string | number
  height?: string | number
  responsive?: boolean
  ratio?: string
}

export const makeStyles = ({ width = '100%', height = '100%' }: VideoPlayerStyleProps) => {
  return {
    containerStyles: css`
      max-width: ${breakpoints.medium};
      & .video-player {
      }
    `,
    playerStyles: css`
      width: ${width};
      height: ${height};
    `,
  }
}
