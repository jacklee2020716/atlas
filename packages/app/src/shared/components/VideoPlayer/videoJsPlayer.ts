import videojs, { VideoJsPlayer, VideoJsPlayerOptions } from 'video.js'
import { RefObject, useEffect, useRef, useState } from 'react'
import 'video.js/dist/video-js.css'

export type VideoJsConfig = {
  src: string
  width?: number
  height?: number
  fluid?: boolean
  fill?: boolean
}

type VideoJsPlayerHook = (config: VideoJsConfig) => [VideoJsPlayer | null, RefObject<HTMLVideoElement>]
export const useVideoJsPlayer: VideoJsPlayerHook = ({ fill, fluid, height, src, width }) => {
  const playerRef = useRef<HTMLVideoElement>(null)
  const [player, setPlayer] = useState<VideoJsPlayer | null>(null)

  useEffect(() => {
    const videoJsOptions: VideoJsPlayerOptions = {
      controls: true,
    }

    const playerInstance = videojs(playerRef.current, videoJsOptions)
    setPlayer(playerInstance)

    return () => {
      playerInstance.dispose()
    }
  }, [])

  useEffect(() => {
    if (!player) {
      return
    }

    player.src({
      src,
      type: 'video/mp4',
    })
  }, [player, src])

  useEffect(() => {
    if (!player || !width) {
      return
    }

    player.width(width)
  }, [player, width])

  useEffect(() => {
    if (!player || !height) {
      return
    }

    player.height(height)
  }, [player, height])

  useEffect(() => {
    if (!player) {
      return
    }

    player.fluid(Boolean(fluid))
  }, [player, fluid])

  useEffect(() => {
    if (!player) {
      return
    }

    // @ts-ignore @types/video.js is outdated and doesn't provide types for some newer video.js features
    player.fill(Boolean(fill))
  }, [player, fill])

  return [player, playerRef]
}
