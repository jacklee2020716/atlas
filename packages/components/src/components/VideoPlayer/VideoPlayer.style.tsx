import { css } from "@emotion/core";
import { breakpoints } from "../../theme";

export type VideoPlayerStyleProps = {
	width?: string | number;
	height?: string | number;
	responsive?: boolean;
	ratio?: string;
};

export let makeStyles = ({
	width = "100%",
	height = "100%",
	responsive = true,
	ratio = "16:9",
}: VideoPlayerStyleProps) => {
	let ratioPerc = ratio
		.split(":")
		.map(x => Number(x))
		.reduce((x, y) => (y / x) * 100);

	return {
		containerStyles: css`
			max-width: ${breakpoints.medium};
			& .video-player {
			}
		`,
		playerStyles: css`
			width: ${width};
			height: ${height};
		`,
	};
};
