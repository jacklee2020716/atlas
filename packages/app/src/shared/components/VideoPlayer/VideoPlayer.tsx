import React from 'react'
import ReactPlayer from 'react-player'
import { VideoPlayerStyleProps, makeStyles } from './VideoPlayer.style'

export type VideoPlayerProps = {
  src?: string
  playing?: boolean
  poster?: string
  controls?: boolean
  volume?: number
  loop?: boolean
  autoPlay?: boolean
  muted?: boolean
  className?: string
  onReady?(): void
  onStart?(): void
  onPlay?(): void
  onPause?(): void
  onBuffer?(): void
  onEnded?(): void
  onError?(error: any): void
  onDuration?(duration: number): void
  onProgress?(state: { played: number; loaded: number }): void
} & VideoPlayerStyleProps

export default function VideoPlayer({
  src,
  poster,
  playing,
  autoPlay,
  loop = false,
  onStart,
  onReady,
  onPlay,
  onBuffer,
  onEnded,
  onDuration,
  onProgress,
  className,
  controls = true,
  ...styleProps
}: VideoPlayerProps) {
  const { playerStyles, containerStyles } = makeStyles(styleProps)
  return (
    <div css={containerStyles}>
      <ReactPlayer
        css={playerStyles}
        width={styleProps.responsive ? '100%' : styleProps.width}
        height={styleProps.responsive ? '100%' : styleProps.height}
        url={src}
        autoPlay={autoPlay}
        light={poster || true}
        className={className}
        playing={playing}
        loop={loop}
        controls={controls}
        onStart={onStart}
        onPlay={onPlay}
        onBuffer={onBuffer}
        onReady={onReady}
        onEnded={onEnded}
        onDuration={onDuration}
        onProgress={onProgress}
        config={{
          file: {
            attributes: {
              className: 'video-player',
            },
          },
        }}
      />
    </div>
  )
}
