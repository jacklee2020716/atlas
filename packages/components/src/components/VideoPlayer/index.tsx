import { memo } from "react"
import VideoPlayer from "./VideoPlayer"
export default memo(VideoPlayer)
