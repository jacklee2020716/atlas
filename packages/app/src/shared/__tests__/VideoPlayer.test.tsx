import React from 'react'
import { mount } from 'enzyme'
import VideoPlayer from '@/shared/components/VideoPlayer'
describe('VideoPlayer component', () => {
  it('Should render Video Player correctly', () => {
    expect(mount(<VideoPlayer src="" />)).toBeDefined()
  })
})
