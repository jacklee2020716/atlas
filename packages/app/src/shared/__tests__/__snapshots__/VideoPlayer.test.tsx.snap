// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer component Should render Video Player correctly 1`] = `ReactWrapper {}`;
